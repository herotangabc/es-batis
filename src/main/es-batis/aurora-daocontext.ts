// import { rdsDataService } from '../aws-sdk'
// import * as uuid from "uuid";
// import type Connection from "mysql2/typings/mysql/lib/Connection";
// import mysql from 'mysql2'
// import { DaoContext } from "./daocontext";

// import init, { InitOptionsReturnType } from 'data-api-client'

// import { /*SecretsManager, */RDSDataService, AWSError } from 'aws-sdk'
// import { preSynthesis, formatResults, ExcuteResponse } from './rds-result-format'

// let secretArn: string = ''

// async function getClusterArnByName(clusterName: string): Promise<RDS.DBClusterMessage> {
//   return new Promise<RDS.DBClusterMessage>((r, j) => {
//     const rds = new RDS();
//     rds.describeDBClusters({ DBClusterIdentifier: clusterName }, (err, data) => {
//       if (err) {
//         j(err)
//       } else {
//         r(data)
//       }
//     })
//   })
// }
// async function getSecretArnByName(secretIdOrName: string): Promise<SecretsManager.DescribeSecretResponse> {
//   return new Promise<SecretsManager.DescribeSecretResponse>((r, j) => {
//     const secretsManager = new SecretsManager();
//     secretsManager.describeSecret({ SecretId: secretIdOrName }, (err, data) => {
//       if (err) {
//         j(err)
//       } else {
//         r(data)
//       }
//     })
//   })
// }


// export default class AuroraDaoContext implements DaoContext {
//   private id: string | undefined;
//   // public loading:boolean|undefined;
//   private dataClient!: RDSDataService
//   private transactionId: string[]
//   // private clusterArn!: string;
//   // private secretArn!: string;
//   private config!: any
//   constructor() {
//     this.id = uuid.v4();
//     this.transactionId = []
//   }
//   async initialize(): Promise<void> {
//     //     if (!clusterArn) {
//     //       const clusterInfo = await getClusterArnByName(process.env.RDS_ARN as string)
//     //       if (clusterInfo?.DBClusters?.[0].EngineMode != 'serverless') {
//     //         throw new Error(`DBCluster Name:${process.env.RDS_NAME} ,only [serverless] RDS type is supported:${clusterInfo?.DBClusters?.[0].EngineMode}
//     // see https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/RDSDataService.html for help`)
//     //       }
//     // this.clusterArn = process.env.RDS_ARN as string
//     // }
//     // if (!secretArn) {
//     //   const secretInfo = await getSecretArnByName(process.env.SECRET_NAME as string)

//     //   secretArn = secretInfo.ARN as string
//     // }
//     // this.clusterArn = clusterArn
//     // this.secretArn = secretArn
//     this.dataClient = rdsDataService
//     this.config = {
//       secretArn: process.env.SECRET_ARN as string,//secretArn,
//       resourceArn: process.env.RDS_ARN as string,
//       database: process.env.DATABASE as string,

//       // includeResultMetadata: true
//     }
//   }
//   public executeStatement(sql: string, parameters: Record<string, any> | Record<string, any>[], transactionId: string | undefined, callback: (err: any, rows: ReturnType<typeof formatResults> | undefined, transactionId: string | undefined) => void): any {
//     const _indx: number = this.find(transactionId, "excute", false);
//     if (_indx >= 0) {
//       transactionId = this.transactionId[_indx]
//     }
//     let isBatch: boolean, params: any, hydrateColumnNames: boolean, args: any, formatOptions: any;
//     try {
//       const preParam = preSynthesis([sql, parameters], { ...this.config, hydrateColumnNames: true, formatOptions: { deserializeDate: true, treatAsLocalDate: true } })
//       isBatch = preParam.isBatch;
//       params = preParam.params;
//       hydrateColumnNames = preParam.hydrateColumnNames;
//       args = preParam.args;
//       formatOptions = preParam.formatOptions;
//     } catch (_err) {
//       callback(_err, undefined, transactionId);
//       return;
//     }
//     const _callback = (err: AWSError, data: ExcuteResponse) => {
//       if (err) {
//         callback(err, undefined, transactionId)
//       } else {
//         const result = formatResults(data, hydrateColumnNames, args[0].includeResultMetadata, formatOptions)
//         console.log(JSON.stringify(data))
//         callback(undefined, result, transactionId)
//       }
//     }

//     //return this.dataClient.query(sql, parameters)
//     // try { // attempt to run the query
//     transactionId ? params.transactionId = transactionId : undefined;
//     // Capture the result for debugging
//     isBatch ? this.dataClient.batchExecuteStatement(params, _callback)
//       : this.dataClient.executeStatement(params, _callback)
//   }

//   public async beginTransaction(): Promise<string | undefined> {
//     return new Promise<string | undefined>((r, j) => {
//       this.dataClient.beginTransaction(this.config as RDSDataService.BeginTransactionRequest, (err, data) => {
//         if (err) {
//           j(err)
//         } else {
//           this.transactionId.push(data.transactionId as string)
//           console.log(`transaction[${data.transactionId}] - begin transaction`)
//           r(data.transactionId)
//         }
//       })
//     })
//   }
//   public async release(): Promise<void> {
//     if (this.transactionId.length > 0) {
//       this.transactionId.forEach(v => {
//         console.error(`transaction[${v}] rollback due to not completed.`)
//         this.rollback(v)
//       })
//       throw new Error(`there's some transactions not handle completed!`)
//     }

//   }

//   public async commit(transactionId: string | undefined = undefined): Promise<string | undefined> {
//     const _indx: number = this.find(transactionId, "commit");
//     transactionId = this.transactionId[_indx]
//     return new Promise<string | undefined>((r, j) => {
//       this.dataClient.commitTransaction({
//         transactionId: transactionId as string,
//         secretArn: this.config.secretArn,
//         resourceArn: this.config.resourceArn,
//       } as RDSDataService.CommitTransactionRequest, (err, data) => {
//         if (err) {
//           //this.transactionId = transactionId
//           j(err)
//         } else {
//           console.log(`transaction[${transactionId}] - commit transaction`)
//           this.transactionId.splice(_indx, 1)
//           r(data.transactionStatus);
//         }
//       });
//     })

//   }
//   private find(transactionId: string | undefined, actionName: 'commit' | 'rollback' | 'excute', required: boolean = true) {
//     if (required && this.transactionId.length === 0) {
//       console.log(`transaction[${transactionId}] - no transaction to ${actionName}.`);
//       return -1;
//     }
//     if (transactionId) {
//       const _indx = this.transactionId.indexOf(transactionId)
//       if (_indx < 0) {
//         console.log(`transaction[${transactionId}] - no transaction to ${actionName}.`);
//       }
//       return _indx
//     } else {
//       return this.transactionId.length - 1;
//     }
//   }

//   public rollback(transactionId: string | undefined = undefined): Promise<string | undefined> {
//     const _indx: number = this.find(transactionId, 'rollback');
//     if (_indx >= 0) {

//       transactionId = this.transactionId[_indx]
//       return new Promise<string | undefined>((r, j) => {
//         this.dataClient.rollbackTransaction({
//           transactionId: transactionId as string,
//           secretArn: this.config.secretArn,
//           resourceArn: this.config.resourceArn,
//         } as RDSDataService.RollbackTransactionRequest, ((err, data) => {
//           this.transactionId.pop()
//           if (err) {
//             j(err)
//           } else {
//             console.log(`transaction[${transactionId}] - rollback transaction`)
//             this.transactionId.splice(_indx, 1)
//             r(data.transactionStatus)
//           }
//         }));
//       })
//     } else {
//       return Promise.resolve(undefined)
//     }
//   }
// }