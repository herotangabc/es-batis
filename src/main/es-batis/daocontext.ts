import { formatResults } from "./rds-result-format";

export type CallbackFnType = (err: any, rows: ReturnType<typeof formatResults> | undefined, transactionId: string | undefined) => void
export interface DaoContext {
  initialize(): Promise<void>;
  beginTransaction(): Promise<string | undefined>;
  release(): Promise<void>;
  executeStatement(sql: string, parameters: any, transactionId: string | undefined, callback: CallbackFnType): void;
  commit(transactionId: string | undefined): Promise<string | undefined>;
  rollback(transactionId: string | undefined): Promise<string | undefined>;
}