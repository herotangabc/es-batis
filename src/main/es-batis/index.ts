import sqlTemplate, { initContext as initDaoContext } from './mapping'
import MySqlDaoContext from './mysql-daocontext'
export { DaoContext } from './daocontext'
export default sqlTemplate
export { initDaoContext, MySqlDaoContext }
export { AuroraEntity, EntityFieldsType } from './orm-base/AuroraEntity'
export { DBColumn } from './orm-base/annotation'