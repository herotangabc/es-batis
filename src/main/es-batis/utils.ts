import { parseExpression } from '@babel/parser'
import { DOMParser } from '@xmldom/xmldom'
import { variables } from 'eval-estree-expression'
import dayjs from 'dayjs'
export { escape, format } from 'mysql2'

export type TestFunction = (...data: any[]) => any

export interface ExtentionDataType {
  readonly payload: Record<string, any>
  // readonly meta:{
  //     preIndexName:string
  //     preItemVarName:string
  // }
}

export const parseDocFromString = (xmlstring: string): Document => {
  let level: string = '',
    message: string = ''
  const doc = new DOMParser({
    locator: {},
    errorHandler: (lev, mess) => {
      level = lev
      message = mess
    },
  }).parseFromString(xmlstring)
  if (level == 'error' || level == 'fatalError') {
    throw new Error(message)
  }
  return doc
}

export type ReturnTypeOfAppendSql = [appended: boolean, extentionData: ExtentionDataType | undefined]

export const buildTestFunction = (testExpression: string): [TestFunction, string[]] => {
  try {
    //获取变量列表并去重
    const vars = variables(parseExpression(testExpression))
    return [new Function(...vars, "'use strict';return " + testExpression) as TestFunction, vars]
  } catch (err) {
    throw new Error(`test expression [${testExpression}] has an error:${(err as Error).message}`)
  }
}

export const lazyFreezeClone = (paramValue: any): any => {
  if (
    paramValue == null ||
    typeof paramValue === 'number' ||
    typeof paramValue === 'boolean' ||
    typeof paramValue === 'string' ||
    paramValue instanceof Date
  ) {
    return paramValue
  } else {
    return Object.freeze({ ...paramValue })
  }
}

/**
 *
 * for each trimStrArray trim only once on head/Tail.
 * ex:
 *  for 'abc','def' in trimStrArray,it will only trim at most once of them
 * on head or on tail,if 'abc' is trimmed on head,'def' will not be trimmed even if
 * it happend on the tail of being trimmed by 'abc' string,
 * but it('def') can be trimmed on the head if it happend on head
 *
 * @param source
 * @param trimStrArray all the different string to trim,if word like,trim as word,other simple trim just as is
 * @param headOrTail 1:head,2:tail,3:both
 */
export const trimHeadOrTail = (
  source: string | undefined,
  trimStrArray: string[] | null,
  headOrTail: 1 | 2 | 3
): string | undefined => {
  if (!trimStrArray || !source || !source.trim().length) {
    return source
  }
  let resultString = source
  let result = 0x00 //01:trim happend on head,02:trim happend on tail,03:trim happend on both
  const wordRegex = /^\w+$/i
  for (const v of trimStrArray) {
    //if v is a word like
    if ((result & 0x01 || !(headOrTail & 0x01)) && (result & 0x02 || !(headOrTail & 0x02))) {
      break
    }
    const wordLike = wordRegex.test(v)
    const trimStringUpper = v.toUpperCase()
    const upperResultStr = resultString.toUpperCase()
    let indexTmp
    if (headOrTail & 0x01 && !(result & 0x01)) {
      //head
      if ((indexTmp = upperResultStr.trimStart().indexOf(trimStringUpper)) == 0) {
        const index = upperResultStr.indexOf(trimStringUpper)
        let flg = true
        if (wordLike) {
          if (
            index + trimStringUpper.length < upperResultStr.length &&
            wordRegex.test(upperResultStr.substr(index + trimStringUpper.length, 1))
          ) {
            flg = false
          }
        }
        if (flg) {
          resultString = resultString.substr(0, index) + resultString.substring(index + trimStringUpper.length)
          result |= 0x01
        }
      }
    }
    if (headOrTail & 0x02 && !(result & 0x02)) {
      //tail
      const upperResultStrTrimend = upperResultStr.trimEnd()
      const indexTmp = upperResultStrTrimend.lastIndexOf(trimStringUpper)
      if (indexTmp + trimStringUpper.length == upperResultStrTrimend.length) {
        let flg = true
        if (wordLike) {
          if (indexTmp != 0 && wordRegex.test(upperResultStr.substr(indexTmp - 1, 1))) {
            flg = false
          }
        }
        if (flg) {
          resultString = resultString.substr(0, indexTmp) + resultString.substring(indexTmp + trimStringUpper.length)
          result |= 0x02
        }
      }
    }
  }

  return resultString
}

export const buildParameters = (
  testFuncParams: string[],
  data: any,
  extentionData: ExtentionDataType | undefined
): any[] => {
  const paramArray: any[] = []
  testFuncParams?.forEach((paraName) => {
    if (paraName == '_') {
      //lazy freeze
      paramArray.push(lazyFreezeClone(data))
    } else if (paraName == '$env') {
      paramArray.push(lazyFreezeClone(process.env))
    } else if (paraName == 'escapeMYSQLLike') {
      paramArray.push(escapeMYSQLLike)
    } else if (extentionData?.payload && paraName in extentionData?.payload) {
      paramArray.push(lazyFreezeClone(extentionData?.payload[paraName]))
    } else {
      if (data == null) {
        throw new Error(`tag variable [${paraName}] value can be found since input parameter is falsy`)
      } else {
        paramArray.push(lazyFreezeClone(data[paraName]))
      }
    }
  })
  return paramArray
}

/**
 * suitable for mysql  LIKE sql,for Sql Server,please refer to
 * https://docs.microsoft.com/en-us/sql/t-sql/language-elements/like-transact-sql?view=sql-server-ver15
 * for example
 * <pre>
 *   <select id="findMatterLikeName">
 *       select MATTER_NAME as matterName from INCIDENT_MATTER_MST
 *       <where>
 *       <bind name="matterNameLike" value="escapeMYSQLLike(matterName)">
 *       AND MATTER_NAME LIKE concat('%',#{matterNameLike},'%')
 *       </where>
 *   </select>
 * </pre>
 * @param escapeStr
 * @returns
 */
export function escapeMYSQLLike(escapeStr: string | undefined | null): string {
  if (!escapeStr) {
    return ''
  }
  return escapeStr.replace(/([%_\\])/g, '\\$1')
}

export function parseDateTime(value: string): Date | undefined {
  if (/^\d{4}-\d{2}-\d{2}[ T]\d{2}:\d{2}:\d{2}(\.\d{1,3})?(Z|(\+|-)\d{2}:?\d{2})?$/.test(value)) {
    const ret = Date.parse(value)
    if (isFinite(ret)) {
      return new Date(ret)
    }
  }
  return undefined
}


/**
 * convert xxxYyyZzz to XXX_YYY_ZZZ
 * @param camelcaseName
 * @returns
 */
export function camelcase2Underline(camelcaseName: string): string {
  if (!camelcaseName) {
    return camelcaseName
  } else if (camelcaseName.indexOf('_') >= 0) {
    return camelcaseName.toUpperCase()
  } else {
    // return camelcaseName.toUpperCase().replace(/(?!\b)([A-Z])/g, (v) => "_" + v);
    return camelcaseName.replace(/\B([A-Z])/g, '_$1').toUpperCase()
  }
}

/**
 * convert XXX_YYY_ZZZ to xxxYyyZzz
 * @param underlineName
 * @returns
 */
export function underline2Camelcase2(underlineName: string): string {
  if (!underlineName) {
    return underlineName
  } else {
    return underlineName.toLowerCase().replace(/(_([a-z]))/g, (v) => v.substring(1, 2).toUpperCase())
  }
}

export function getCurrentTime(): string {
  return dayjs().format('YYYY-MM-DD HH:mm:ss.SSS')
}