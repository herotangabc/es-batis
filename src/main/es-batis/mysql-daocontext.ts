// import { SecretsManager } from 'aws-sdk'
import { Connection, createConnection } from 'mysql2'
import * as uuid from 'uuid'
import { CallbackFnType, DaoContext } from './daocontext'

// async function getSecretValueBySecretArn(secretIdOrArn: string): Promise<SecretsManager.GetSecretValueResponse> {
//   return new Promise<SecretsManager.GetSecretValueResponse>((r, j) => {
//     const secretsManager = new SecretsManager()
//     secretsManager.getSecretValue({ SecretId: secretIdOrArn }, (err, data) => {
//       if (err) {
//         j(err)
//       } else {
//         r(data)
//       }
//     })
//   })
// }

type MySQLConnectionInfo = {
  charset?: string
  host: string
  user?: string
  password?: string
  database?: string
  debug?: string
  connectTimeout: number
}


class MySqlDaoContext implements DaoContext {
  private id: string | undefined
  private connection!: Connection
  private secretKeyValue!: Record<string, string> | undefined
  #conectionInfo?: MySQLConnectionInfo
  constructor(conectionInfo?: MySQLConnectionInfo) {
    this.#conectionInfo = conectionInfo;
    this.id = uuid.v4()
  }
  async initialize(): Promise<void> {
    this.connection = createConnection({
      charset: process.env.charset ?? 'utf8',
      host: process.env.host,
      user: process.env.user,
      database: process.env.database,
      password: process.env.password,
      debug: process.env.debug == 'true',
      ...this.#conectionInfo,
      namedPlaceholders: true
    })

    // if (!this.secretKeyValue) {
    //   console.log('SECRET_ARN::::::', process.env.SECRET_ARN)
    //   const secretInfo = await getSecretValueBySecretArn(process.env.SECRET_ARN as string)
    //   console.log('secretInfo::::::', secretInfo)
    //   this.secretKeyValue = JSON.parse(secretInfo?.SecretString as string)
    // }

    // this.connection = createConnection({
    //   // charset: process.env.charset ?? 'utf8',
    //   host: this.secretKeyValue?.host,
    //   user: this.secretKeyValue?.username,
    //   database: process.env.DATABASE,
    //   password: this.secretKeyValue?.password,
    //   // debug: process.env.debug == 'true',
    //   // connectTimeout: 60000, //60 second
    //   // namedPlaceholders: true,
    // })
  }
  public executeStatement(
    sql: string,
    parameters: any,
    transactionId: string | undefined,
    callback: CallbackFnType
  ): void {
    this.connection.execute(sql, parameters, (err, data, fields) => {
      if (err) {
        callback(err, undefined, undefined)
      } else {
        if (Array.isArray(data)) {
          callback(
            err,
            {
              records: data as any[],
            },
            undefined
          )
        } else {
          callback(
            err,
            {
              numberOfRecordsUpdated: data.affectedRows as number,
              insertId: data.insertId as number,
            },
            undefined
          )
        }
      }
    })
  }

  public async beginTransaction(): Promise<string | undefined> {
    return new Promise<string | undefined>((r, j) => {
      this.connection.beginTransaction((err) => {
        if (err) {
          j(err)
        } else {
          r(undefined)
        }
      })
    })
  }
  public async release(): Promise<void> {
    if (this.connection) {
      this.connection.destroy()
    }
  }
  public async commit(): Promise<string | undefined> {
    return new Promise<string | undefined>((r, j) => {
      this.connection.commit((err) => {
        if (err) {
          j(err)
        } else {
          r(undefined)
        }
      })
    })
  }
  public rollback(): Promise<string | undefined> {
    return new Promise<string | undefined>((r, j) => {
      try {
        this.connection.rollback(() => {
          r(undefined)
        })
      } catch (err) {
        j(`${err}`)
      }
    })
  }
}
export default MySqlDaoContext
