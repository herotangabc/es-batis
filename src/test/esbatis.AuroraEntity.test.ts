import sqlTemplate, { DaoContext, initDaoContext, AuroraEntity, DBColumn } from 'es-batis';
import { templateManager, SqlCommand } from 'es-batis/mapping'
import { camelcase2Underline } from 'es-batis/utils'
import type { SqlBound } from 'es-batis/mapping'
//fix Encoding not recognized: 'cesu8' (searched as: 'cesu8') issue
//https://stackoverflow.com/questions/46227783/encoding-not-recognized-in-jest-js
// require('mysql2/node_modules/iconv-lite').encodingExists('foo');
import dotenv from 'dotenv'

import dayjs from 'dayjs'
import { FieldPacket, QueryError } from 'mysql2';
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
dayjs.extend(timezone)
dayjs.extend(utc)
// import mysql  from 'mysql2'

// import type {Connection} from 'mysql2'

let result: { records?: Record<string, any>, numberOfRecordsUpdated?: number } | null = null;
let _sql: string | null = null
let _parameters: any = null;

class SystemMst extends AuroraEntity {
  @DBColumn({ idType: 'uniqueIndex' })
  public systemId!: string | null
  @DBColumn({ idType: 'uniqueIndex' })
  public systemName: string | null = null
}

class CodeMst extends AuroraEntity {
  @DBColumn({ idType: 'uniqueIndex' })
  public code!: string | null
  @DBColumn({ idType: 'uniqueIndex' })
  public name: string | null = null
  public order: number | null = null
}

class MockDaoContext implements DaoContext {
  initialize(): Promise<void> {
    return Promise.resolve();
  }
  beginTransaction(): Promise<string | undefined> {
    return Promise.resolve(undefined);
  }
  executeStatement(sql: string, parameters: any, transactionid: string, callback: (err: any, rows: any, _fields: any) => void) {
    _sql = sql
    _parameters = parameters
    callback(undefined, result, [])
  }
  release(): Promise<void> {
    return Promise.resolve();
  }
  commit(): Promise<string | undefined> {
    return Promise.resolve(undefined);
  }
  rollback(): Promise<string | undefined> {
    return Promise.resolve(undefined);
  }
}

beforeAll(async () => {
  dotenv.config()
  dayjs.tz.setDefault('Etc/UCT')
  // const diff = dayjs.tz('2022-01-01', "UTC").diff(dayjs.tz('2022-01-01', dayjs.tz.guess()))
  // const currentTZ = dayjs.tz.guess()
  // connection = mysql.createConnection({
  //     host: process.env.DB_HOST,
  //     user: process.env.DB_USER,
  //     database: process.env.DB_NAME,
  //     password: process.env.DB_PASSWORD,
  //     charset:'utf8'
  // })
  const dao = new MockDaoContext()
  initDaoContext({ dao })
});

beforeEach(async () => {
  templateManager.clear()
})

afterAll(async () => {
  execute: (sql: string, values: any, callback?: (
    err: QueryError | null,
    result: any,
    fields: FieldPacket[]
  ) => any) => {

  }


  // connection.destroy()
})

test('1. entity composed id test', async () => {
  expect(sqlTemplate).not.toBeUndefined()
  const systemMst: SystemMst = new SystemMst()
  systemMst.systemId = '123'
  systemMst.systemName = '456'
  result = { records: [] }
  await systemMst.load();
  expect(_sql).toMatchSnapshot()
  expect(_parameters).toMatchSnapshot()
}, 5000)

test('2. entity composed id test with one record', async () => {
  expect(sqlTemplate).not.toBeUndefined()
  const systemMst: SystemMst = new SystemMst()
  systemMst.systemId = '123'
  systemMst.systemName = '456'
  result = {
    records: [{
      systemId: '123',
      systemName: '456',
      updateDate: '2022-12-07 12:00:00.235'
    }]
  }
  await systemMst.load();
  expect(_sql).toMatchSnapshot()
  expect(_parameters).toMatchSnapshot()
}, 5000)

test('3. entity load with more than one record throws error', async () => {
  expect(sqlTemplate).not.toBeUndefined()
  const systemMst: SystemMst = new SystemMst()
  systemMst.systemId = '123'
  systemMst.systemName = '456'
  result = {
    records: [{
      systemId: '123',
      systemName: '456',
      updateDate: '2022-12-07 12:00:00.235'
    }, {
      systemId: '234',
      systemName: '657',
      updateDate: '2022-12-08 12:00:00.235'
    }]
  }
  let f = async () => {
    await systemMst.load(undefined, ['systemId'], true);
  }
  expect(f).rejects.toThrowError()
}, 5000)

test('4. entity save test', async () => {
  expect(sqlTemplate).not.toBeUndefined()
  const systemMst: SystemMst = new SystemMst()
  systemMst.systemId = '123'
  systemMst.systemName = '456'
  result = {
    numberOfRecordsUpdated: 1
  }
  await systemMst.save()
  expect(_sql).toMatchSnapshot()
  expect(_parameters).toMatchSnapshot()
}, 5000)

test('5. entity save test', async () => {
  expect(sqlTemplate).not.toBeUndefined()
  const codeMst: CodeMst = new CodeMst()
  codeMst.code = '123'
  codeMst.name = '456'
  result = {
    numberOfRecordsUpdated: 1
  }
  await codeMst.save()
  expect(_sql).toMatchSnapshot()
  expect(_parameters).toMatchSnapshot()
}, 5000)