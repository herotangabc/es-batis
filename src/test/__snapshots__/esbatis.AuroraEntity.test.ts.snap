// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. entity composed id test 1`] = `
"SELECT 
SYSTEM_NAME as systemName,
SYSTEM_ID as systemId 
FROM SYSTEM_MST
 WHERE SYSTEM_ID=:systemId AND SYSTEM_NAME=:systemName"
`;

exports[`1. entity composed id test 2`] = `
Object {
  "systemId": "123",
  "systemName": "456",
}
`;

exports[`2. entity composed id test with one record 1`] = `
"SELECT 
SYSTEM_NAME as systemName,
SYSTEM_ID as systemId 
FROM SYSTEM_MST
 WHERE SYSTEM_ID=:systemId AND SYSTEM_NAME=:systemName"
`;

exports[`2. entity composed id test with one record 2`] = `
Object {
  "systemId": "123",
  "systemName": "456",
}
`;

exports[`4. entity save test 1`] = `
"INSERT INTO SYSTEM_MST 
(SYSTEM_NAME,
SYSTEM_ID)
VALUES(:systemName,
:systemId)"
`;

exports[`4. entity save test 2`] = `
Object {
  "systemId": "123",
  "systemName": "456",
}
`;

exports[`5. entity save test 1`] = `
"INSERT INTO CODE_MST 
(NAME,
ORDER,
CODE)
VALUES(:name,
:order,
:code)
ON DUPLICATE KEY UPDATE
ORDER = :order"
`;

exports[`5. entity save test 2`] = `
Object {
  "code": "123",
  "name": "456",
  "order": null,
}
`;
